### PR TITLE
Aggregate vision data

### LINKED ISSUE #230: Aggregate vision data in the network_input node
**Is your feature request or enhancement related to a problem? What specifically could be better? Please describe.**
The `network_input` node should aggregate the camera data into a full snapshot of the field, and publish data for the entire field, rather than for each camera individually.

**Describe the solution you'd like**
Aggregate the received data into Ball/Team objects in the `network_input` node. This way we can update only parts of the data as we receive single camera frames. Then each time a frame is received we can publish the entire Ball/Team data message again. Timestamps should be tracked with the data.

**Describe alternatives you've considered**
None

**Additional context**
None

### PR DESCRIPTION
**Pull request by MathewMacDougall**
### Description

Aggregate vision data in the `network_input` node so that we can publish a complete "world state" rather than fractions of the field at a time.

### Testing Done

No tests added, all existing tests pass. Manually verified messages are being published, and the positions/velocities are still reported correctly.

### Resolved Issues

resolves #230 

### Length Justification

It was close to 500 but I forgot to remove the `ros_message` utils from the network_input node in #328 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [x] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom`
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Comment by MathewMacDougall** [comment]
This PR is still slightly WIP but I wanted to let people start taking a look.

There are issues when the robots/ball pass over camera boundaries. For example, robot velocities temporarily become 0. This is likely because multiple cameras are reporting on the time/location of the robot but this data is being handled/filtered separately. I'll do more debugging in the morning.

File: src/thunderbots/software/network_input/filter/robot_filter.h [hunk 1]
```diff
@@ -19,7 +19,7 @@ typedef struct SSLRobotData_t
     Angle orientation;
     double confidence;
     Timestamp timestamp;
-} SSLRobotData;
+} SSLRobotDetection;
```
Comment (garethellis0) [inline src/thunderbots/software/network_input/filter/robot_filter.h:22]: Does this need to be public?

**Review by garethellis0 — CHANGES_REQUESTED** [review]
This is really nice! Pretty clean. :)

File: src/thunderbots/software/network_input/backend.h [hunk 2]
```diff
@@ -32,19 +34,18 @@ class Backend
...
 
     /**
      * Given a new protobuf packet, returns a Field message containing the most up to date
      * Field geometry
      *
-     * @param packet The SSL Vision packet containing new data
+     * @param packet The SSL_GeometryFieldSize packet containing new field data
      *
      * @return a Field message containing the most up to date Field geometry. If the
      * packet does not contain any new Field geometry, returns std::nullopt
```
[the remark this answers is not in the document]
Comment (MathewMacDougall) [reply src/thunderbots/software/network_input/backend.h]: No, comments not all updated yet.

File: src/thunderbots/software/network_input/filter/ball_filter.h [hunk 3]
```diff
@@ -15,20 +16,8 @@ typedef struct
...
- * something like std::pair so it's very explicit what data is being accessed, since
- * it's accessed by name rather than by index (first/second)
- */
-typedef struct FilteredBallData_t
-{
-    Point position;
-    Vector velocity;
     Timestamp timestamp;
-} FilteredBallData;
+} SSLBallDetection;
```
[the remark this answers is not in the document]
Comment (MathewMacDougall) [reply src/thunderbots/software/network_input/filter/ball_filter.h:20]: Yes it should be public, `SSLBallDetections` are constructed in the backend before being passed to the filter. This abstracts the filter from needing to know about the protobuf packet structure/data.

File: src/thunderbots/software/network_input/backend.h [hunk 4]
```diff
@@ -55,8 +56,7 @@ class Backend
      * @return a Team message containing the most up to date filtered friendly team data.
      * If the packet does not contain any new Team or Robot data, returns std::nullopt
      */
-    std::optional<thunderbots_msgs::Team> getFilteredFriendlyTeamMsg(
-        const SSL_WrapperPacket &packet);
+    Team getFilteredFriendlyTeamData(const SSL_WrapperPacket &packet);
```
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/thunderbots/software/network_input/backend.h]: Whoopsie! Sorry, I'm blind.

File: src/thunderbots/software/network_input/filter/robot_filter.h (hunk 1 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/network_input/filter/robot_filter.h:22]: Yes it should be public, `SSLRobotDetections` are constructed in the backend before being passed to the filter. This abstracts the filter from needing to know about the protobuf packet structure/data.

File: src/thunderbots/software/network_input/backend.cpp [hunk 5]
```diff
@@ -1,149 +1,224 @@
...
+    // RightGoalTopLine
+    // RightGoalBottomLine
+    // RightGoalDepthLine
+    // LeftGoalTopLine
+    // LeftGoalBottomLine
+    // LeftGoalDepthLine
+    // LeftFieldLeftPenaltyStretch
+    // LeftFieldRightPenaltyStretch
+    // RightFieldLeftPenaltyStretch
+    // RightFieldRightPenaltyStretch
```
[the remark this answers is not in the document]
Comment (MathewMacDougall) [reply src/thunderbots/software/network_input/backend.cpp:77]: I agree it's not ideal, but I don't think an enum works in this case. These field lines/circles are only distinguishable by their `name` field:

```
  // Name of this field marking.
  required string name = 1;
```

And there's no explicit guarantee the shapes will be published in the same order allowing us to index them with an enum. So we determined the strings through observation/testing and have that list in the comments to it's easy to access other lines/shapes in the future.

[the remark this answers is not in the document]
Comment (garethellis0) [reply src/thunderbots/software/network_input/backend.cpp:77]: Hmm. Could we at least populate a struct with the lines we need and throw an exception/warn/skip the current iteration if we're missing some?

[the remark this answers is not in the document]
Comment (MathewMacDougall) [reply src/thunderbots/software/network_input/backend.cpp:77]: The struct seems like an almost pointless middelman here. 

I can check for the existence of the keys and catch/throw any exceptions due to missing keys (although we'll already get an exception if the key isn't found). We wouldn't want to skip the iteration in this case, since it would be slightly messy to propagate an optional all the way back up to the main loop, and because this kind of error indicates we are not handling the protocol correctly which is not something we should tolerate.

[the remark this answers is not in the document]
Comment (garethellis0) [reply src/thunderbots/software/network_input/backend.cpp:77]: Good argument. I'm fine with leaving this as-is then, I imagine any exception a missing key generates should be pretty easy to trace.

**Comment by MathewMacDougall** [comment]
tl;dr of my last commit:
* Aggregated the detection packets for each camera, and pass all this data to the filters at once. It's up to the filters to pick/discard the data they want
    * This fixes the "0 velocity between cameras" issue from before
* Tweaks to the backend to take vectors of proto packets
* Comments updated

File: src/thunderbots/software/network_input/backend.cpp [hunk 6]
```diff
@@ -1,149 +1,218 @@
...
-std::optional<thunderbots_msgs::Team> Backend::getFilteredFriendlyTeamMsg(
-    const SSL_WrapperPacket &packet)
+Team Backend::getFilteredFriendlyTeamData(std::vector<SSL_DetectionFrame> detections)
 {
-    if (packet.has_detection())
-    {
-        const SSL_DetectionFrame &detection                = packet.detection();
-        std::vector<SSLRobotData> friendly_team_robot_data = std::vector<SSLRobotData>();
+    std::vector<SSLRobotDetection> friendly_robot_detections =
+        std::vector<SSLRobotDetection>();
```
Comment (garethellis0) [inline src/thunderbots/software/network_input/backend.cpp]: nit; but I'd avoid duplicate type declarations by making this 
```
auto friendly_robot_detections = std::vector<SSLRobotDetection>();
```

File: src/thunderbots/software/network_input/main.cpp [hunk 7]
```diff
@@ -83,55 +87,91 @@ int main(int argc, char** argv)
...
-            {
-                friendly_team_publisher.publish(*friendly_team_msg);
+                auto detection = packet.detection();
+                // Add the detection to the map, replacing any existing values
+                auto ret = latest_detection_data.insert(
+                    std::make_pair(detection.camera_id(), detection));
+                if (!ret.second)
+                {
+                    ret.first->second = detection;
+                }
```
Comment (garethellis0) [inline src/thunderbots/software/network_input/main.cpp:119]: I'm a bit confused about what `ret` is here and what we're doing with it?

File: src/thunderbots/software/network_input/networking/ssl_vision_client.cpp [hunk 8]
```diff
@@ -46,13 +46,13 @@ void SSLVisionClient::handleDataReception(const boost::system::error_code& error
...
         // several times to process several packets. This is why we immediately store any
-        // received and processed data into the packet_queue. If we didn't save the
+        // received and processed data into the packet_vector. If we didn't save the
         // received data somewhere else, the data would be overwritten by the packet
         // being handled in the next handleDataReception function call
         auto packet_data = SSL_WrapperPacket();
         packet_data.ParseFromArray(raw_received_data_.data(),
                                    static_cast<int>(num_bytes_received));
-        packet_queue.emplace(packet_data);
+        packet_vector.emplace_back(packet_data);
```
Comment (garethellis0) [inline src/thunderbots/software/network_input/networking/ssl_vision_client.cpp:55]: Hmm, why the switch over to `vector`? It was kinda nice to use a structure that explicitly preserves the order of insertion? Perhaps `std::deque` might be a better choice?

File: src/thunderbots/software/network_input/main.cpp (hunk 7 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/network_input/main.cpp:119]: the return type of map::insert is a pair, where the first element is an iterator pointing to the object in the map, and the second is a boolean if the item was inserted. If the item was not inserted, it means the key already existed and the iterator points to that entry. So then we modify that entry to replace the item.

tl;dr used to add entries to the map and allow replacement of existing items

File: src/thunderbots/software/network_input/networking/ssl_vision_client.cpp (hunk 8 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/network_input/networking/ssl_vision_client.cpp:55]: Order doesn't _really_ matter here though. I can leave it as `queue`, `vector` was just nicer to iterator over.

Comment (garethellis0) [reply src/thunderbots/software/network_input/networking/ssl_vision_client.cpp:55]: Ah yeah, I guess because all the packets are timestamped.

File: src/thunderbots/software/network_input/main.cpp (hunk 7 above)
Comment (garethellis0) [reply src/thunderbots/software/network_input/main.cpp:119]: Ok, could we add a comment or two just explaining that in this section? Just some really brief ones, like `// Check that we inserted succesfully` would be fine.

File: src/thunderbots/software/network_input/networking/ssl_vision_client.cpp (hunk 8 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/network_input/networking/ssl_vision_client.cpp:55]: There's also no guarantee (that I'm aware of) that we enqueue packets in the order they were received by the network. I'm not sure if the `handle...` functions are called in the order they were set up.

File: src/thunderbots/software/network_input/backend.cpp (hunk 6 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/network_input/backend.cpp]: fixed

File: src/thunderbots/software/network_input/main.cpp (hunk 7 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/network_input/main.cpp:119]: fixed

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada:
